persistent-postgresql/Database/Persist/Postgresql.hs

```diff
@@ -1712,16 +1712,70 @@ upsertWhere
--- It will insert a record if no matching unique key exists.
--- If a unique key exists, it will update the relevant field with a user-supplied value, however,
--- it will only do this update on a user-supplied condition.
--- For example, here's how this method could be called like such:
+-- | Postgres specific 'upsertManyWhere'.
 --
--- upsertManyWhere [record] [recordField =. newValue] [recordField !=. newValue]
+-- The first argument determines a list of new records to insert.
 --
--- Called thusly, this method will insert a new record (if none exists) OR update a recordField with a new value
--- assuming the condition in the last block is met.
+-- If there's unique key collisions for some or all of the proposed insertions,
+-- you can use the second argument to specify which fields, under which
+-- conditions (using 'HandleUpdateCollision' strategies,) will be copied from
+-- each proposed, conflicting new record onto its corresponding row already present
+-- in the database. Helpers such as 'copyField', 'copyUnlessEq',
+-- 'copyUnlessNull' and 'copyUnlessEmpty' are exported from this module to help
+-- with this, as well as the constructors for 'HandleUpdateCollision'.
+-- For example, you may have a patch of updates to apply, some of which are NULL
+-- to represent "no change"; you'd use 'copyUnlessNull' for each applicable
+-- field in this case so any new values in the patch which are NULL don't
+-- accidentally overwrite existing values in the database which are non-NULL.
+--
+-- Further updates to the matching records already in the database can be
+-- specified in the third argument, these are arbitrary updates independent of
+-- the new records proposed for insertion. For example, you can use this to
+-- update a timestamp or a counter for all conflicting rows.
+--
+-- You can use the fourth argument to provide arbitrary conditions to constrain
+-- the above updates. This way, you can choose to only alter a subset of the
+-- conflicting existing rows while leaving the others alone. For example, you
+-- may want to copy fields from new records and update a timestamp only when the
+-- corresponding existing row hasn't been soft-deleted.
+--
+--
+-- For example,
+--
+-- @
+--   upsertManyWhere
+--     [record] -- new records to insert if there's no conflicts
+--     [ copyField recordField1 -- for each conflicting existing row, replace the value of recordField1 with the one present in the conflicting new record
+--     , copyUnlessEq recordField2 -- only replace the existing value if it's different from the one present in the conflicting new record
+--     , copyUnlessNull recordField3 -- only replace the existing value if the new value is non-NULL (i.e. don't replace existing values with NULLs.)
+--     ]
+--     [recordField4 =. arbitraryValue] -- update recordField4 with an arbitrary new value
+--     [recordField4 !=. anotherValue] -- only apply the above updates for conflicting rows that meet this condition
+-- @
```

Comment:
This looks a little noisy - can we try and get the line length below 80 columns? Putting the comments directly above the item in question

```suggestion
-- @
--   upsertManyWhere
--     [record]                         -- (1) 
--     [ copyField recordField1         -- (2) 
--     , copyUnlessEq recordField2      -- (3) 
--     , copyUnlessNull recordField3    -- (4) 
--     ]
--     [recordField4 =. arbitraryValue] -- (5) 
--     [recordField4 !=. anotherValue]  -- (6) 
-- @
-- 
-- 1. new records to insert if there's no conflicts
-- 2. for each conflicting existing row, replace the value of recordField1 with the one present in the conflicting new record
-- 3. only replace the existing value if it's different from the one present in the conflicting new record
-- 4. only replace the existing value if the new value is non-NULL (i.e. don't replace existing values with NULLs.)
-- 
-- 5. update recordField4 with an arbitrary new value
-- 6. only apply the above updates for conflicting rows that meet this condition
```

Reply:
It may also be helpful to refer to the example you gave above, where you don't want to hit something with a soft-delete.

Possibly,

```haskell
mkPersist sqlSettings [persistLowerCase|
  Product
    name         Text
    stock        Int Maybe
    price        Int Maybe
    updateCount  Int
    deletedAt    UTCTime Maybe

    UniqueProduct name
|]

do
  products <- loadCSV
  upsertManyWhere 
    products
    [ copyUnlessNull ProductStock
    , copyField ProductPrice
    ]
    [ updatedCount +=. 1 ]
    [ ProductDeletedAt !=. Nothing ]
```